src/authorization/permissions/DeleteParentExtractor.ts

```diff
@@ -0,0 +1,44 @@
+import type { Operation } from '../../http/Operation';
+import type { ResourceSet } from '../../storage/ResourceSet';
+import type { IdentifierStrategy } from '../../util/identifiers/IdentifierStrategy';
+import { ModesExtractor } from './ModesExtractor';
+import type { AccessMap } from './Permissions';
+import { AccessMode } from './Permissions';
+
+/**
+ * In case a resource is being deleted but does not exist,
+ * the server response code depends on the access modes the agent has on the parent container.
+ * In case the agent has read access on the parent container, a 404 should be returned,
+ * otherwise it should be 401/403.
+ *
+ * This class adds support for this by making sure the agent has read access on the parent container
+ * in case the target resource does not exist.
+ */
+export class DeleteParentExtractor extends ModesExtractor {
+  private readonly source: ModesExtractor;
+  private readonly resourceSet: ResourceSet;
+  private readonly identifierStrategy: IdentifierStrategy;
+
+  public constructor(source: ModesExtractor, resourceSet: ResourceSet, identifierStrategy: IdentifierStrategy) {
```

Comment:
```suggestion
  public constructor(source: ModesExtractor, store: ResourceSet, identifierStrategy: IdentifierStrategy) {
```

(suggestion, just because I wondered what the second parameter was)

Reply:
Going to keep `resourceSet` for consistency with other classes that do something similar. `ResourceSet` is actually an interface name you suggested when the `hasResource` function was added 😄 .

Reply:
I like the interface name; naming the variable after the interface is like `string: String` 🙂 I'm looking of a hint of _what_ kind of resource set we are talking about here.

Reply:
Well what we want is an object that can tell us if a certain resource exists. I don't think `store` make this clearer. A `ResourceSet` specifically is an object that can give us that information. Similar to the parameter right next to it actually: `identifierStrategy: IdentifierStrategy`, which is an object that can perform certain identifier related operations. `string: String` can be a valid input parameter in specific situations, for example a stringbuffer that concatenates strings.